GitUI/CommandsDialogs/RevisionDiff.cs

```diff
@@ -279,19 +279,27 @@ private void DiffContextMenu_Opening(object sender, CancelEventArgs e)
             if (selectedRevisions.Count > 1)
                 artificialRevSelected = artificialRevSelected || selectedRevisions[selectedRevisions.Count - 1].IsArtificial();
 
+            //Many options have no meaning for artificial commits or submodules
+            //Hide the obviously no action options when single selected, handle them in actions if multi select
+
             // disable items that need exactly one selected item
             bool isExactlyOneItemSelected = DiffFiles.SelectedItems.Count() == 1;
             var isCombinedDiff = isExactlyOneItemSelected &&
                 DiffFiles.CombinedDiff.Text == DiffFiles.SelectedItemParent;
             var isAnyCombinedDiff = DiffFiles.SelectedItemParents.Any(item => item == DiffFiles.CombinedDiff.Text);
-            var enabled = isExactlyOneItemSelected && !isCombinedDiff;
+
             openWithDifftoolToolStripMenuItem.Enabled = !isAnyCombinedDiff;
-            saveAsToolStripMenuItem1.Enabled = enabled;
-            cherryPickSelectedDiffFileToolStripMenuItem.Enabled = enabled;
-            diffShowInFileTreeToolStripMenuItem.Enabled = isExactlyOneItemSelected;
-            fileHistoryDiffToolstripMenuItem.Enabled = isExactlyOneItemSelected;
-            blameToolStripMenuItem.Enabled = isExactlyOneItemSelected;
-            resetFileToToolStripMenuItem.Enabled = !isCombinedDiff;
+            saveAsToolStripMenuItem1.Visible = !isCombinedDiff && isExactlyOneItemSelected && !DiffFiles.SelectedItem.IsSubmodule;
+            cherryPickSelectedDiffFileToolStripMenuItem.Visible = !isCombinedDiff && isExactlyOneItemSelected &&
+                !(DiffFiles.SelectedItem.IsSubmodule || selectedRevisions[0].Guid == GitRevision.UnstagedGuid ||
+                (DiffFiles.SelectedItem.IsNew || DiffFiles.SelectedItem.IsDeleted) && selectedRevisions[0].Guid == GitRevision.IndexGuid);
+            //Visibility of FileTree is not known, assume (CommitInfoTabControl.Contains(TreeTabPage);)
+            diffShowInFileTreeToolStripMenuItem.Visible = isExactlyOneItemSelected && !selectedRevisions[0].IsArtificial(); 
+            fileHistoryDiffToolstripMenuItem.Enabled = isExactlyOneItemSelected && !(DiffFiles.SelectedItem.IsNew && selectedRevisions[0].IsArtificial());
+            blameToolStripMenuItem.Enabled = isExactlyOneItemSelected && !(DiffFiles.SelectedItem.IsSubmodule || selectedRevisions[0].IsArtificial());
+            resetFileToToolStripMenuItem.Enabled = !isCombinedDiff &&
+                !(isExactlyOneItemSelected &&
+                (DiffFiles.SelectedItem.IsSubmodule || DiffFiles.SelectedItem.IsNew) && selectedRevisions[0].Guid == GitRevision.UnstagedGuid);
```

Comment:
I'm sorry but I find these conditions too complex for comprehension and maintenance.
Would you consider creating a `RevisionDiffController` which would have methods something like: 

```
bool CanSaveAs(IList<GitRevision> selectedRevisions, IEnumerable<GitItemStatus> selectedFiles, bool isCombinedDiff);
bool CanCherryPick(IList<GitRevision> selectedRevisions, IEnumerable<GitItemStatus> selectedFiles, bool isCombinedDiff);
// etc
```

this way the logic can be tested and is much more maintainable.
The `DiffContextMenu_Opening` implementation becomes pretty simple too:

```
            // disable items that need exactly one selected item
            bool isExactlyOneItemSelected = DiffFiles.SelectedItems.Count() == 1;
            var isCombinedDiff = isExactlyOneItemSelected && DiffFiles.CombinedDiff.Text == DiffFiles.SelectedItemParent;
            var isAnyCombinedDiff = DiffFiles.SelectedItemParents.Any(item => item == DiffFiles.CombinedDiff.Text);

            openWithDifftoolToolStripMenuItem.Enabled = !isAnyCombinedDiff;
            saveAsToolStripMenuItem1.Visible = _revisionDiffController.CanSave(selectedRevisions, DiffFiles.SelectedItems, isCombinedDiff);
            cherryPickSelectedDiffFileToolStripMenuItem.Visible = _revisionDiffController.CanCherryPick(selectedRevisions, DiffFiles.SelectedItems, isCombinedDiff);
            // etc
```

Reply:
#4106, #4089, #4087, #4092 (merged though) involves the same file and I would like to make this refactoring only once. Maybe similar for #4084 and a few other merged too (in other files).

@Russkie: Can you approve the PRs (with comments), then I will make a new PR including the RevisionDiff commits?

Reply:
ok. merging on a condition this will get reworked

Reply:
Thanks. Can I stack #4087 on top of #4089 (or opposite) and make the refactoring there?